api/paidAction/zap.js

```diff
@@ -0,0 +1,151 @@
+import { USER_ID } from '@/lib/constants'
+import { msatsToSats, satsToMsats } from '@/lib/format'
+import { notifyZapped } from '@/lib/webPush'
+
+export const anonable = true
+export const supportsPessimism = true
+export const supportsOptimism = true
+
+export async function getCost ({ sats }) {
+  return satsToMsats(sats)
+}
+
+export async function perform ({ invoiceId, sats, id: itemId, ...args }, { me, cost, tx }) {
+  const feeMsats = cost / BigInt(100)
+  const zapMsats = cost - feeMsats
+  itemId = parseInt(itemId)
+
+  let invoiceData = {}
+  if (invoiceId) {
+    invoiceData = { invoiceId, invoiceActionState: 'PENDING' }
+    // store a reference to the item in the invoice
+    await tx.invoice.update({
+      where: { id: invoiceId },
+      data: { actionId: itemId }
+    })
+  }
+
+  const acts = await tx.itemAct.createManyAndReturn({
+    data: [
+      { msats: feeMsats, itemId, userId: me?.id || USER_ID.anon, act: 'FEE', ...invoiceData },
+      { msats: zapMsats, itemId, userId: me?.id || USER_ID.anon, act: 'TIP', ...invoiceData }
+    ]
+  })
+
+  const [{ path }] = await tx.$queryRaw`
+    SELECT ltree2text(path) as path FROM "Item" WHERE id = ${itemId}::INTEGER`
+  return { id: itemId, sats, act: 'TIP', path, actIds: acts.map(act => act.id) }
```

Comment:
We need `actIds` to correlate `perform` and `onPaid` absent an `invoiceId`, e.g. paying with fee credits.

Reply:
Mutating input parameters is an anti-pattern because it's a side effect and side effects are hidden from the calling context. Introducing side effects to functions mean that when we change the calling context, we would have to know about all the potential side effects of the functions it calls, wherever they are, because they can't be seen from the calling context.